### DISCUSSION TITLE
Simple example on how to use Excalidraw in standalone (html/js) with Backend on PHP Server

### DISCUSSION THREAD

**Discussion by gitubbi** [excalidraw/excalidraw#6569] [Q&A]
Hello,

Here is a simple exemple of a standalone use of Excalidraw in html/js with backend for a php server.

- At start, Excali is loaded with your custom library file, that should be named "libray.excalidrawlib" in the same folder as the script.

- The main menu is changed to save & export on the server (instead of on the device).

- It allows to save jsonfile.excalidraw or export png file with embbeded scene metadata, that allows to re-edit the scene later.

- Add also some custom configuration of Excalidraw UI for example.

**Usage:** 

Rename the "index.txt" in "index.php" and put this file in a folder of your server's document root, then call it from client with the syntax:

`https://myhostname/pathto/index.php`

or to re-edit a previous file, use query string:  ?name=filename  (file should be in the same folder as php script)

`https://myhostname/pathto/index.php?name=image.png (or jsonfile.excalidraw)`

The files are saved in the script directory to simplify this example.

For a png image, Excalidraw doesn't load the image itself, but reload the complete Scene, extracting it from the metadata embedded inside, so you can continue to modify it again.
 
[index.txt](https://github.com/excalidraw/excalidraw/files/11423744/index.txt)

**Comment by marcelfolaron** [comment]
Thank you! This was super helpful! 
You don't by any chance know how to change the platform url for real time collaboration?

**Comment by gitubbi** [comment]
Very glad it could help you. I have done it as there is only few examples in docs to show the use of excalidraw without react environment that I don't know and it is difficult to guess how to adapt from react to js.

I don't use collaborative mode so unfortunately i can't help you on this side, but i remember to have read some times ago that you can't use collaborative mode on self-hosted version without react or something like that, but i'm not sure about that. 

All the properties you can change are described in the API docs here, maybe you will find what you need:

[https://docs.excalidraw.com/docs/@excalidraw/excalidraw/api](https://docs.excalidraw.com/docs/@excalidraw/excalidraw/api)

**Reply by marcelfolaron** [reply]
Yeah it appears the the socket urls are passed in as environment variables but since I am only using the npm package I am not running the docker image. Couldn't find a way to inject a new url. Will do more digging and see if I can find a workaround or build my own collab integration.

**Reply by aman811** [reply]
Hi, I'm new to this and looking to do what you suggested. Can you tell me what's the custom library file supposed to be? In your code, you are already including excalidraw.production.min.js . I thought that was all that had to included as per integration guide?

**Comment by haukefrahmann** [comment]
on the website I found this >> "The Ref support has been removed in v0.17.0 so if you are using refs, please update the integration to use the excalidrawAPI."


How do intergrate the api with the new version?
I get always an error on save.

**Reply by LDSmj** [reply]
have you fixed the error on the save function??

**Reply by haukefrahmann** [reply]
Yes, a friend of mine got me there:

const options =  {
          langCode: 'de-DE',
          name: name,
          // ref: (api) => { setExcalidrawAPI(api); },
          initialData: {
              libraryItems:    library.libraryItems,
              elements:        scene.elements,
              files:           scene.files,
              appState:        { ...scene.appState, ...appStateDefault() },
          }
       };


       const App = () => {

         [excalidrawAPI, setExcalidrawAPI] = React.useState(null);

          // return React.createElement(
          //           React.Fragment, null,
          //             React.createElement( 'div', { style: { height: "100vh" } },
          //               React.createElement( ExcalidrawLib.Excalidraw, options,
          //                  React.createElement( ExcalidrawLib.MainMenu, {}, MainMenu() ),
          //               ),
          //             ),
          //           );
           return React.createElement(
               React.Fragment, null,
               React.createElement('div', { style: { height: "100vh" } },
                   React.createElement(ExcalidrawLib.Excalidraw,
                       {
                           ...options,
                           excalidrawAPI: api => setExcalidrawAPI(api)
                       }, React.createElement(ExcalidrawLib.MainMenu, {}, MainMenu()) ),
               ),
           );
       };

**Reply by LDSmj** [reply]
Thank you very much! It has taken me almost a whole day just to find how to fix the error.

**Reply by aman811** [reply]
Hi, can you tell me what's the custom library file is supposed to be, in the suggested code?

 $library  = 'libray.excalidrawlib';   //Library Folder
    $asset    = $dir . 'Excalidraw/';     //Excali Assets Folder (locale & fonts)
    $pathFile = __DIR__ . '/' ;           //Files Folder (for json & image)


I thought all I needed to include was excalidraw.production.min.js, as suggested in the integration guide? Any help would be appreciated. What folder do I need to assign to $library?

**Comment by aman811** [comment]
When I try to load excalidraw file, it doesn't work. Images work fine though. It seems the file generated via export is not correct, because if I feed another file created via excalidraw website, it works fine. Any ideas?

**Reply by aman811** [reply]
For anyone who faces the same issue, the excalidraw is adding this line in app state, which causes the issue.
    "collaborators": {},
